numpy/lib/_stride_tricks_impl.py

```diff
@@ -101,7 +102,10 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
     array = np.asarray(DummyArray(interface, base=x))
     # The route via `__interface__` does not preserve structured
     # dtypes. Since dtype should remain unchanged, we set it explicitly.
-    array.dtype = x.dtype
+    with warnings.catch_warnings():
```

Comment:
I'd prefer not to have `_set_dtype` at all, but if we keep it, then this seems a good place to use it.

Reply:
Agreed, the idea of having this work around, is exactly the ability to avoid any warning context managers both in NumPy (and potentially outside of NumPy).

Reply:
Agreed. I wanted to be on the safe side since there is a difference between `.dtype = ...` and `._set_dtype(...)` for subclasses that override setting attributes. Here array is created by `np.asarray` so it is fine and I will update the code.